clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -797,7 +797,8 @@ static void VisitField(CXXRecordDecl *Owner, RangeTy &&Item, QualType ItemTy,
 template <typename RangeTy, typename... Handlers>
 static void VisitArrayElements(RangeTy Item, QualType FieldTy,
                                Handlers &... handlers) {
-  const ConstantArrayType *CAT = cast<ConstantArrayType>(FieldTy);
+  const ConstantArrayType *CAT =
```

Comment:
Instead of casts throughout this, you want to be doing getas.  https://clang.llvm.org/doxygen/classclang_1_1Type.html#adae68e1f4c85ede2d36da45fbefc48a2

This will make sure you get the unqualified desugared type.

Reply:
I'm not sure if getAs works with ArrayType. While working on another patch I hit this assert -https://clang.llvm.org/doxygen/Type_8h_source.html#l07153

This was the code that failed:
      if (const ConstantArrayType *CAT = FieldTy->getAs<ConstantArrayType>())

Reply:
Ah, looks like there is a specific one for ArrayType: https://clang.llvm.org/doxygen/classclang_1_1ASTContext.html#a54d2386e12cd16689dd812d63b85560f

Reply:
Yea I was just looking into that. Do you know if that will work when checking elaborated types? It wasn't clear to me by looking at definition.

Reply:
I think so?  It does some canonicalization by calling getSplitDesugaredType, so that would imply to me that it removes typedef-type sugaring.

Reply:
Never mind. I was looking at the wrong definition. I think it will work. This is what the comment says - 

     If the type is an instance of the specified class,
     /// return the Type pointer for the underlying maximally pretty type.  This
     /// is a member of ASTContext because this may need to do some amount of
     /// canonicalization, e.g. to move type qualifiers into the element type.